### PR TITLE
Implement the UC Minimum Income Floor

### PR DESCRIPTION
**Pull request by nikhilwoodruff**
For the self-employed, there's a [minimum income floor](https://www.gov.uk/government/publications/universal-credit-and-self-employment-quick-guide/universal-credit-and-self-employment-quick-guide) that they're treated as earning, which has previously not been modeled in openfisca-uk. I've added most of it here (just need to double check everything), along with the minimum wage variable and parameters which were necessary.

@MaxGhenis FYI - pretty much done, just needs a double check (tests added for all variables).

### CODE REVIEW

File: openfisca_uk/variables/finance/benefit/family/universal_credit.py [hunk 1]
```diff
@@ -506,7 +506,18 @@ def formula(benunit, period, parameters):
             "self_employment_income",
             "miscellaneous_income",
         ]
-        earned_income = aggr(benunit, period, INCOME_COMPONENTS)
+        personal_earned_income = add(
```
Comment (MaxGhenis) [inline openfisca_uk/variables/finance/benefit/family/universal_credit.py:509]: Does `add` aggregate over `benunit.members`? Trying to parse what this section is doing. A comment or two would be helpful.

Would this be clearer if some of it was in a person-level variable?

File: openfisca_uk/variables/finance/benefit/family/universal_credit.py [hunk 2]
```diff
@@ -506,7 +506,18 @@ def formula(benunit, period, parameters):
...
+            benunit.members, period, INCOME_COMPONENTS
+        )
+        personal_mif_capped_income = where(
+            benunit.members("UC_MIF_applies", period),
+            max_(
```
Comment (MaxGhenis) [inline openfisca_uk/variables/finance/benefit/family/universal_credit.py:514]: Maybe make this a separate variable (Python, not OpenFisca)

File: openfisca_uk/variables/finance/benefit/family/universal_credit.py [hunk 3]
```diff
@@ -568,3 +579,50 @@ def formula(benunit, period, parameters):
             * benunit("claims_UC", period)
             * benunit("is_UC_eligible", period)
         )
+
+
+class UC_MIF_applies(Variable):
```
Comment (MaxGhenis) [inline openfisca_uk/variables/finance/benefit/family/universal_credit.py:584]: Should there also be a ['start-up period'](https://www.gov.uk/government/publications/universal-credit-and-self-employment-quick-guide/universal-credit-and-self-employment-quick-guide#the-start-up-period) variable, set to `False`?

File: openfisca_uk/variables/finance/benefit/family/universal_credit.py [hunk 4]
```diff
@@ -568,3 +579,50 @@ def formula(benunit, period, parameters):
...
+    reference = (
+        "https://www.legislation.gov.uk/uksi/2013/376/regulation/62/2021-04-06"
+    )
+    definition_period = YEAR
+
+    def formula(person, period, parameters):
+        return person("self_employment_income", period) > 0
+
+
+class UC_expected_hours_per_week(Variable):
```
Comment (MaxGhenis) [inline openfisca_uk/variables/finance/benefit/family/universal_credit.py:598]: Do we have any other variables like this that are just parameters? Seems unnecessary and to set a burdensome precedent.

File: openfisca_uk/variables/finance/income.py [hunk 5]
```diff
@@ -241,3 +241,58 @@ class base_net_income(Variable):
     entity = Person
     label = "Existing net income for the person to use as a base in microsimulation"
     definition_period = YEAR
+
+
+class is_apprentice(Variable):
```
Comment (MaxGhenis) [inline openfisca_uk/variables/finance/income.py:246]: Should this have a false default, or is false the default default?

File: openfisca_uk/variables/finance/income.py [hunk 6]
```diff
@@ -241,3 +241,58 @@ class base_net_income(Variable):
...
+    label = "Minimum wage category"
+    definition_period = YEAR
+
+    def formula(person, period, parameters):
+        age = person("age", period)
+        return select(
+            [
+                person("is_apprentice", period),
+                age < 18,
+                (age >= 18) & (age <= 20),
```
Comment (MaxGhenis) [inline openfisca_uk/variables/finance/income.py:275]: I thought there was a way to use something like `between` for this?

**Review by MaxGhenis — COMMENTED** [review]
Good stuff, how does this affect total UC?

File: openfisca_uk/variables/finance/benefit/family/universal_credit.py (hunk 1 above)
Comment (nikhilwoodruff) [reply openfisca_uk/variables/finance/benefit/family/universal_credit.py:509]: I'll add some comments - but I think the Minimum Income Floor is on a per-person basis, with couples' MIF floors being added together (but only if they are both self-employed). This does a lot of the person-level calculation, before summing to the benefit unit level.

File: openfisca_uk/variables/finance/benefit/family/universal_credit.py (hunk 4 above)
Comment (nikhilwoodruff) [reply openfisca_uk/variables/finance/benefit/family/universal_credit.py:598]: I agree, but I wasn't sure yet of the exact circumstances for reduced expected hours of work, will revisit once it's clear

File: openfisca_uk/variables/finance/income.py (hunk 6 above)
Comment (nikhilwoodruff) [reply openfisca_uk/variables/finance/income.py:275]: True, will use that instead

File: openfisca_uk/parameters/benefit/universal_credit/work_requirements/default_expected_hours.yaml [hunk 7]
```diff
@@ -0,0 +1,9 @@
+description: Default expected hours worked per week
+values:
+  2013-04-29:
+    value: 35
+metadata:
+  unit: hour
+  reference:
+    - title: The Universal Credit Regulations 2013, section 88
+      href: https://www.legislation.gov.uk/uksi/2013/376/regulation/88/2013-04-29
```
[the remark this answers is not in the document]
Comment (nikhilwoodruff) [reply openfisca_uk/parameters/benefit/universal_credit/work_requirements/default_expected_hours.yaml]: Added

File: openfisca_uk/variables/finance/income.py (hunk 6 above)
Comment (nikhilwoodruff) [reply openfisca_uk/variables/finance/income.py:275]: Oh, turns out I need to add `between` to openfisca-tools.

File: openfisca_uk/tests/policy/baseline/finance/income/minimum_wage.yaml [hunk 8]
```diff
@@ -0,0 +1,7 @@
+- name: Minimum wage for a person over 25
```
[the remark this answers is not in the document]
Comment (nikhilwoodruff) [reply openfisca_uk/tests/policy/baseline/finance/income/minimum_wage.yaml:1]: Added

File: openfisca_uk/tests/policy/baseline/finance/benefit/family/universal_credit/UC_MIF_applies.yaml [hunk 9]
```diff
@@ -0,0 +1,22 @@
+- name: No self-employment income
+  period: 2020
+  absolute_error_margin: 0
+  input:
+    employment_income: 1
```
[the remark this answers is not in the document]
Comment (nikhilwoodruff) [reply openfisca_uk/tests/policy/baseline/finance/benefit/family/universal_credit/UC_MIF_applies.yaml:5]: By default, it's zero, so I've removed the entry

File: openfisca_uk/variables/finance/benefit/family/universal_credit.py (hunk 2 above)
Comment (nikhilwoodruff) [reply openfisca_uk/variables/finance/benefit/family/universal_credit.py]: Done

File: openfisca_uk/variables/finance/benefit/family/universal_credit.py (hunk 3 above)
Comment (nikhilwoodruff) [reply openfisca_uk/variables/finance/benefit/family/universal_credit.py:570]: Added

File: openfisca_uk/variables/finance/benefit/family/universal_credit.py (hunk 4 above)
Comment (nikhilwoodruff) [reply openfisca_uk/variables/finance/benefit/family/universal_credit.py]: Moved

File: openfisca_uk/variables/finance/income.py (hunk 5 above)
Comment (nikhilwoodruff) [reply openfisca_uk/variables/finance/income.py:246]: Technically zero is the default default, but it'll be cast to False. But I think we should specify in the abscence of a formula

**Pull request merged**